Import telegraf in nestjs app

Hi all. I`m created a simple module (telegram-bot) in nestjs app.
Here is a simple service importing telegraf - telegram-bot.service.ts:

```
import { Injectable, OnApplicationBootstrap } from '@nestjs/common';
import { Telegraf } from 'telegraf';
import config from 'config';

interface TelegramConfig {
  token: string
}

const TELEGRAM_CONFIG: TelegramConfig = config.get('telegram');

@Injectable()
export class TelegramBotService implements OnApplicationBootstrap {
  private bot: Telegraf;

  onApplicationBootstrap(): void {
    const { token } = TELEGRAM_CONFIG;
    this.bot = new Telegraf(token);
    this.bot.start((ctx) => ctx.reply('Hello'));
    this.bot.launch();
  }
}
```
But compilation errors appear in console:

```
node_modules/telegraf/typings/composer.d.ts:30:25 - error TS1005: ']' expected.

30     [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>;
                           ~~

node_modules/telegraf/typings/composer.d.ts:30:28 - error TS1005: ';' expected.

30     [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>;
                              ~~~~~~~~~~~~~~~~

node_modules/telegraf/typings/composer.d.ts:30:47 - error TS1005: '(' expected.

30     [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>;
                                                 ~

node_modules/telegraf/typings/composer.d.ts:30:48 - error TS1135: Argument expression expected.

30     [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>;
                                                  ~

node_modules/telegraf/typings/composer.d.ts:30:73 - error TS1005: '(' expected.

30     [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>;
                                                                           ~
....
```
What are the solutions?

--- a/src/composer.ts
+++ b/src/composer.ts
@@ -560,7 +560,7 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
   >(
     triggers: ReadonlyArray<(text: string, ctx: C) => RegExpExecArray | null>,
     ...fns: MatchedMiddleware<C, T>
-  ): MiddlewareFn<C> {
+  ): MiddlewareFn<MatchedContext<C, T>> {
     const handler = Composer.compose(fns)
     return (ctx, next) => {
       const text =
@@ -570,6 +570,7 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
         ctx.inlineQuery?.query
       if (text === undefined) return next()
       for (const trigger of triggers) {
+        // @ts-expect-error
         const match = trigger(text, ctx)
         if (match) {
           // @ts-expect-error define so far unknown property `match`

--- a/src/composer.ts
+++ b/src/composer.ts
@@ -14,7 +14,6 @@ type Triggers<C> = MaybeArray<
 type Predicate<T> = (t: T) => boolean
 type AsyncPredicate<T> = (t: T) => Promise<boolean>
 
-/** @deprecated Adds `ctx.match`, unlike `MatchedContext`! */
 type MatchedMiddleware<
   C extends Context,
   T extends tt.UpdateType | tt.MessageSubType = 'message' | 'channel_post'
@@ -46,8 +45,7 @@ export type NarrowedContext<
     updateType: UpdateTypes<U>
   }
 /**
- * Maps `Composer.mount`'s `updateType` to a type
- * that narrows down `tt.Update` when intersected with it.
+ * Maps [[`Composer.on`]]'s `updateType` to a `tt.Update` subtype.
  */
 type MountMap = {
   [T in tt.UpdateType]: Extract<tg.Update, Record<T, object>>

--- a/src/composer.ts
+++ b/src/composer.ts
@@ -2,10 +2,8 @@
 
 import * as tg from './core/types/typegram'
 import * as tt from './telegram-types'
-import Context, { GetUpdateContent, UpdateTypes } from './context'
+import Context, { UpdateTypes } from './context'
 import { Middleware, MiddlewareFn, MiddlewareObj } from './middleware'
-import { PropOr } from './deunionize'
-import { SnakeToCamelCase } from './core/helpers/string'
 
 type MaybeArray<T> = T | T[]
 export type MaybePromise<T> = T | Promise<T>
@@ -32,12 +30,6 @@ type MatchedContext<
   T extends tt.UpdateType | tt.MessageSubType
 > = NarrowedContext<C, MountMap[T]>
 
-type Getter<U extends tg.Update, P extends string> = PropOr<
-  GetUpdateContent<U>,
-  P,
-  undefined
->
-
 /**
  * Narrows down `C['update']` (and derived getters)
  * to specific update type `U`.
@@ -45,15 +37,12 @@ type Getter<U extends tg.Update, P extends string> = PropOr<
  * Used by [[`Composer`]],
  * possibly useful for splitting a bot into multiple files.
  */
-export type NarrowedContext<C extends Context, U extends tg.Update> = C &
-  {
-    [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>
-  } & {
-    update: U
+export type NarrowedContext<
+  C extends Context,
+  U extends tg.Update
+> = Context<U> &
+  Omit<C, keyof Context> & {
     updateType: UpdateTypes<U>
-    senderChat: Getter<U, 'sender_chat'>
-    from: Getter<U, 'from'>
-    chat: Getter<U, 'chat'>
   }
 /**
  * Maps `Composer.mount`'s `updateType` to a type

--- a/src/composer.ts
+++ b/src/composer.ts
@@ -14,6 +14,7 @@ type Triggers<C> = MaybeArray<
 type Predicate<T> = (t: T) => boolean
 type AsyncPredicate<T> = (t: T) => Promise<boolean>
 
+/** @deprecated Adds `ctx.match`, unlike `MatchedContext`! */
 type MatchedMiddleware<
   C extends Context,
   T extends tt.UpdateType | tt.MessageSubType = 'message' | 'channel_post'
@@ -113,7 +114,10 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
   /**
    * Registers middleware for handling specified commands.
    */
-  command(command: MaybeArray<string>, ...fns: MatchedMiddleware<C, 'text'>) {
+  command(
+    command: MaybeArray<string>,
+    ...fns: NonemptyReadonlyArray<Middleware<MatchedContext<C, 'text'>>>
+  ) {
     return this.use(Composer.command<C>(command, ...fns))
   }
 
@@ -207,7 +211,11 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
   /**
    * Registers a middleware for handling /start
    */
-  start(...fns: ReadonlyArray<Middleware<C & { startPayload: string }>>) {
+  start(
+    ...fns: NonemptyReadonlyArray<
+      Middleware<MatchedContext<C, 'text'> & { startPayload: string }>
+    >
+  ) {
     const handler = Composer.compose(fns)
     return this.command('start', (ctx, next) => {
       const entity = ctx.message.entities![0]!
@@ -219,14 +227,16 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
   /**
    * Registers a middleware for handling /help
    */
-  help(...fns: NonemptyReadonlyArray<Middleware<C>>) {
+  help(...fns: NonemptyReadonlyArray<Middleware<MatchedContext<C, 'text'>>>) {
     return this.command('help', ...fns)
   }
 
   /**
    * Registers a middleware for handling /settings
    */
-  settings(...fns: NonemptyReadonlyArray<Middleware<C>>) {
+  settings(
+    ...fns: NonemptyReadonlyArray<Middleware<MatchedContext<C, 'text'>>>
+  ) {
     return this.command('settings', ...fns)
   }
 
@@ -599,7 +609,7 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
    */
   static command<C extends Context>(
     command: MaybeArray<string>,
-    ...fns: MatchedMiddleware<C, 'text'>
+    ...fns: NonemptyReadonlyArray<Middleware<MatchedContext<C, 'text'>>>
   ): MiddlewareFn<C> {
     if (fns.length === 0) {
       // @ts-expect-error command is really the middleware

--- a/src/context.ts
+++ b/src/context.ts
@@ -1,8 +1,8 @@
 import * as tg from './core/types/typegram'
 import * as tt from './telegram-types'
+import { Deunionize, PropOr, UnionKeys } from './deunionize'
 import ApiClient from './core/network/client'
 import Telegram from './telegram'
-import { UnionKeys } from './deunionize'
 
 type Tail<T> = T extends [unknown, ...infer U] ? U : never
 
@@ -10,11 +10,11 @@ type Shorthand<FName extends Exclude<keyof Telegram, keyof ApiClient>> = Tail<
   Parameters<Telegram[FName]>
 >
 
-export class Context {
+export class Context<U extends Deunionize<tg.Update> = tg.Update> {
   readonly state: Record<string | symbol, any> = {}
 
   constructor(
-    readonly update: tg.Update,
+    readonly update: U,
     readonly tg: Telegram,
     readonly botInfo: tg.UserFromGetMe
   ) {}
@@ -32,80 +32,70 @@ export class Context {
   }
 
   get message() {
-    if (!('message' in this.update)) return undefined
-    return this.update.message
+    return this.update.message as PropOr<U, 'message'>
   }
 
   get editedMessage() {
-    if (!('edited_message' in this.update)) return undefined
-    return this.update.edited_message
+    return this.update.edited_message as PropOr<U, 'edited_message'>
   }
 
   get inlineQuery() {
-    if (!('inline_query' in this.update)) return undefined
-    return this.update.inline_query
+    return this.update.inline_query as PropOr<U, 'inline_query'>
   }
 
   get shippingQuery() {
-    if (!('shipping_query' in this.update)) return undefined
-    return this.update.shipping_query
+    return this.update.shipping_query as PropOr<U, 'shipping_query'>
   }
 
   get preCheckoutQuery() {
-    if (!('pre_checkout_query' in this.update)) return undefined
-    return this.update.pre_checkout_query
+    return this.update.pre_checkout_query as PropOr<U, 'pre_checkout_query'>
   }
 
   get chosenInlineResult() {
-    if (!('chosen_inline_result' in this.update)) return undefined
-    return this.update.chosen_inline_result
+    return this.update.chosen_inline_result as PropOr<U, 'chosen_inline_result'>
   }
 
   get channelPost() {
-    if (!('channel_post' in this.update)) return undefined
-    return this.update.channel_post
+    return this.update.channel_post as PropOr<U, 'channel_post'>
   }
 
   get editedChannelPost() {
-    if (!('edited_channel_post' in this.update)) return undefined
-    return this.update.edited_channel_post
+    return this.update.edited_channel_post as PropOr<U, 'edited_channel_post'>
   }
 
   get callbackQuery() {
-    if (!('callback_query' in this.update)) return undefined
-    return this.update.callback_query
+    return this.update.callback_query as PropOr<U, 'callback_query'>
   }
 
   get poll() {
-    if (!('poll' in this.update)) return undefined
-    return this.update.poll
+    return this.update.poll as PropOr<U, 'poll'>
   }
 
   get pollAnswer() {
-    if (!('poll_answer' in this.update)) return undefined
-    return this.update.poll_answer
+    return this.update.poll_answer as PropOr<U, 'poll_answer'>
   }
 
   get myChatMember() {
-    if (!('my_chat_member' in this.update)) return undefined
-    return this.update.my_chat_member
+    return this.update.my_chat_member as PropOr<U, 'my_chat_member'>
   }
 
   get chatMember() {
-    if (!('chat_member' in this.update)) return undefined
-    return this.update.chat_member
+    return this.update.chat_member as PropOr<U, 'chat_member'>
   }
 
-  get chat() {
+  get chat(): Getter<U, 'chat'> {
     return (
       this.chatMember ??
       this.myChatMember ??
       getMessageFromAnySource(this)
-    )?.chat
+    )?.chat as Getter<U, 'chat'>
   }
 
   get senderChat() {
-    return getMessageFromAnySource(this)?.sender_chat
+    return getMessageFromAnySource(this)?.sender_chat as Getter<
+      U,
+      'sender_chat'
+    >
   }
 
   get from() {
@@ -118,7 +108,7 @@ export class Context {
       this.chatMember ??
       this.myChatMember ??
       getMessageFromAnySource(this)
-    )?.from
+    )?.from as Getter<U, 'from'>
   }
 
   get inlineMessageId() {
@@ -593,7 +583,12 @@ export type GetUpdateContent<
   ? U['callback_query']['message']
   : U[UpdateTypes<U>]
 
-function getMessageFromAnySource(ctx: Context) {
+type Getter<U extends Deunionize<tg.Update>, P extends string> = PropOr<
+  GetUpdateContent<U>,
+  P
+>
+
+function getMessageFromAnySource<U extends tg.Update>(ctx: Context<U>) {
   return (
     ctx.message ??
     ctx.editedMessage ??

--- a/src/core/helpers/string.ts
+++ b/src/core/helpers/string.ts
@@ -1,5 +1,0 @@
-export type SnakeToCamelCase<
-  S extends string
-> = S extends `${infer T}_${infer U}`
-  ? `${T}${Capitalize<SnakeToCamelCase<U>>}`
-  : `${S}`

--- a/src/deunionize.ts
+++ b/src/deunionize.ts
@@ -1,7 +1,7 @@
 export type PropOr<
   T extends object | undefined,
   P extends string | symbol | number,
-  D
+  D = undefined
 > = T extends Partial<Record<P, unknown>> ? T[P] : D
 
 export type UnionKeys<T> = T extends unknown ? keyof T : never

